sys/src/go/cmd/tmpfs/tmpfs.go

```diff
@@ -0,0 +1,344 @@
+package main
+
+import (
+	"bytes"
+	"flag"
+	"fmt"
+	"log"
+	"net"
+	"os"
+	"sync"
+
+	"harvey-os.org/internal/tmpfs"
+	"harvey-os.org/pkg/ninep"
+	"harvey-os.org/pkg/ninep/protocol"
+)
+
+var (
+	networktype = flag.String("ntype", "tcp4", "Default network type")
+	netaddr     = flag.String("addr", ":5641", "Network address")
+	debug       = flag.Int("debug", 0, "Print debug messages")
+	pkg         = flag.String("package", "", "tar.gz package to open")
+)
+
+type fileServer struct {
+	sync.Mutex
+
+	Versioned bool
+
+	ioUnit  protocol.MaxSize
+	archive *tmpfs.Archive
+	files   map[protocol.FID]*FidEntry
+}
+
+// FidEntry wraps an Entry with the instance data required for a fid reference
+type FidEntry struct {
+	tmpfs.Entry
+
+	// Username to be used for all entries in this hierarchy
+	uname string
+
+	// We can't know how big a serialized dentry is until we serialize it.
+	// At that point it might be too big. We save it here if that happens,
+	// and on the next directory read we start with that.
+	oflow []byte
+
+	// Index of next child to return when reading a directory
+	nextChildIdx int
+}
+
+func newFidEntry(entry tmpfs.Entry, uname string) *FidEntry {
+	return &FidEntry{Entry: entry, uname: uname, oflow: nil, nextChildIdx: -1}
+}
+
+// Rversion initiates the session
+func (fs *fileServer) Rversion(msize protocol.MaxSize, version string) (protocol.MaxSize, string, error) {
+	if version != "9P2000" {
+		return 0, "", fmt.Errorf("%v not supported; only 9P2000", version)
+	}
+	return msize, version, nil
+}
+
+// Rattach attaches a fid to the root for the given user.  aname and afid are not used.
+func (fs *fileServer) Rattach(fid protocol.FID, afid protocol.FID, uname string, aname string) (protocol.QID, error) {
+	if afid != protocol.NOFID {
+		return protocol.QID{}, fmt.Errorf("We don't do auth attach")
```

Comment:
the linux 9p fs code expects very specific strings, and if a return like this does not match, it returns some useless number. Consider sending a message known to linux 9p client, if possible. If not, that's fine too, but I've been burned by this.